File: Euler/pro17.cpp
/*If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?


NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with British usage.*/

#include <iostream>
using std::cout;
using std::endl;
using std::string;

int determine( int );
int oneNUMBER( int );
int twoNUMBER( int );
int threeNUMBER( int );

enum littleNums { ONE = 1, TWO = 3, THREE = 5, FOUR = 4, FIVE = 4, SIX = 3, SEVEN = 5, EIGHT = 5, NINE = 4, TEN = 3, ELEVEN = 6, TWELVE = 6, THIRTEEN = 8, FOURTEEN = 8, FIFTEEN = 7, SIXTEEN = 7, SEVENTEEN = 9, EIGHTEEN = 8, NINETEEN = 8 };

enum bigNums { TWENTY = 6, THIRTY = 6, FORTY = 5, FIFTY = 5, SIXTY = 5, SEVENTY = 7, EIGHTY = 6, NINETY = 6, HUNDRED = 7, THOUSAND = 8 };

int main()
{
   unsigned long long sum = 0; // Resultado del programa
   
   const int limit = 5;
   for( int i = 1; i <= limit; i++ )
   {
       sum += determine( i );
   }
   
   cout << sum << endl;
   
   return 0;
}

int determine( int n )
{
      if( n < 20 )
      return oneNUMBER( n );
      
      if(20 <= n < 100 )
      return twoNUMBER( (n - (n%10)) /10 ) + oneNUMBER( n % 10 );
      
      if( 100 <= n < 1000 )
      return threeNUMBER( (n - (n%100)) / 100) + twoNUMBER( (n - (n%10)) /10 ) + oneNUMBER( n % 10 ) + 3/*AND*/;
      if ( n == 1000)
      return ONE + THOUSAND;
}

int oneNUMBER( int n )
{
   switch(n)
   {
      case 1:
      return ONE;
      
      case 2:
      return TWO;
      
      case 3:
      return THREE;
      
      case 4:
      return FOUR;
      
      case 5:
      return FIVE;
      
      case 6:
      return SIX;
      
      case 7:
      return SEVEN;
      
      case 8:
      return EIGHT;
      
      case 9:
      return NINE;
      
      case 10:
      return TEN;
      
      case 11:
      return ELEVEN;
      
      case 12:
      return TWELVE;
      
      case 13:
      return THIRTEEN;
      
      case 14:
      return FOURTEEN;
      
      case 15:
      return FIFTEEN;
      
      case 16:
      return SIXTEEN;
      
      case 17:
      return SEVENTEEN;
      
      case 18:
      return EIGHTEEN;
      
      case 19:
      return NINETEEN;
   }
}

int twoNUMBER( int n )
{
   switch(n)
   {
      case 2:
      return TWENTY;
      
      case 3:
      return THIRTY;
      
      case 4:
      return FORTY;
      
      case 5:
      return FIFTY;
      
      case 6:
      return SIXTY;
      
      case 7:
      return SEVENTY;
      
      case 8:
      return EIGHTY;
      
      case 9:
      return NINETY;
   }
}

int threeNUMBER( int n )
{
   int num = 0;
   switch(n)
   {
      case 1:
      num = ONE;
      
      case 2:
      num = TWO;
      
      case 3:
      num = THREE;
      
      case 4:
      num = FOUR;
      
      case 5:
      num = FIVE;
      
      case 6:
      num = SIX;
      
      case 7:
      num = SEVEN;
      
      case 8:
      num = EIGHT;
      
      case 9:
      num = NINE;
   }
   
   return num + HUNDRED;
}

File: Euler/pro17.py
# encoding=UTF-8
__author__ = 'Alejandro'
import string

'''If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?


NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with British usage.'''

NUMEROS = {'one': 1,
           'two': 2,
           'three': 3,
           'four': 4,
           'five': 5,
           'six': 6,
           'seven': 7,
           'eight': 8,
           'nine': 9,
           'ten': 10,
           'eleven': 11,
           'twelve': 12,
           'thirteen': 13,
           'fourteen': 14,
           'fifteen': 15,
           'sixteen': 16,
           'seventeen': 17,
           'eighteen': 18,
           'nineteen': 19}

DENOMINACION = {'twenty': 20,
                'thirty': 30,
                'forty': 40,
                'fifty': 50,
                'sixty': 60,
                'seventy': 70,
                'eighty': 80,
                'ninety': 90,
                'hundred': 100,
                'thousand': 1000}


def formar_en_palabras(n):
    assert isinstance(n, int)
    if n > 1000:
        return "ERROR: Mayor que 1000"

    texto = ""
    while n > 0:
        if n == 1000:
            return "one thousand"
        elif 100 <= n < 1000:
            for clave, valor in NUMEROS.items():
                if str(n // 100) == str(valor):
                    texto = clave + ' hundred '
                    n -= (n // 100) * 100
                    break
        elif 20 <= n < 100:
            for clave, valor in DENOMINACION.items():
                if str(n - (n % 10)) == str(valor):
                    if texto != "":
                        texto += 'and '
                    texto += clave
                    n -= (n // 10) * 10
                    if n > 0:
                        texto += '-'
                    break
        elif 1 <= n < 20:
            for clave, valor in NUMEROS.items():
                if n == valor:
                    if texto[len(texto)-8:] == 'hundred ':
                        texto += 'and '
                    texto += clave
                    n = 0
        else:
            print('No deberia llegar hasta aqui')
    return texto


suma = 0
for i in range(1, 1001):
    palabras = formar_en_palabras(i)
    # print(i, palabras)
    for palabra in palabras.split():
        for letra in palabra:
            if letra in string.ascii_lowercase:
                suma += 1

print(suma)